README.md
[![CircleCI](https://dl.circleci.com/status-badge/img/gh/giantswarm/azure-operator/tree/master.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/giantswarm/azure-operator/tree/master)

# azure-operator

The azure-operator manages Kubernetes clusters running in Giantnetes on Azure.

## Branches

- `thiccc`
    - Up to and including version v2.9.0.
    - Contains all versions of legacy controllers (reconciling AzureConfig CRs) up
      to and including v2.9.0.
- `master`
    - From version v3.0.0.
    - Contains only the latest version of controllers.

## Getting Project

Clone the git repository: https://github.com/giantswarm/azure-operator.git

### How to build

Build it using the standard `go build` command.

```
go build github.com/giantswarm/azure-operator
```

## Running azure-operator

Create an azure-operator role using [tenant.tmpl.json](policies/tenant.tmpl.json) role definition (replace SUBSCRIPTION_ID):

```bash
az role definition create --role-definition @tenant.tmpl.json
```

If you have a service proivder you want to reuse add the azure-operator role
(replace CLIENT_ID):

```bash
az role assignment create --assignee ${CLIENT_ID} --role azure-operator
```

Otherwise create a service provider with the azure-operator role (replace
SUBSCRIPTION_ID):

```bash
export CODENAME=cluster1
az ad sp create-for-rbac -n $CODENAME-azure-operator-sp --role="azure-operator" --scopes="/subscriptions/${SUBSCRIPTION_ID}" --years 10
```

Follow [this guide][examples-local].

[examples-local]: https://github.com/giantswarm/azure-operator/blob/master/examples/README.md

## Contact

- Mailing list: [giantswarm](https://groups.google.com/forum/!forum/giantswarm)
- IRC: #[giantswarm](irc://irc.freenode.org:6667/#giantswarm) on freenode.org
- Bugs: [issues](https://github.com/giantswarm/azure-operator/issues)

## Contributing & Reporting Bugs

See [CONTRIBUTING](CONTRIBUTING.md) for details on submitting patches, the
contribution workflow as well as reporting bugs.

## Pre-commit hook

- [Install pre-commit](https://pre-commit.com/#installation).
- [Install the git hook scripts](https://pre-commit.com/#3-install-the-git-hook-scripts)

Now you can [run the pre-commit against all files in the repository](https://pre-commit.com/#4-optional-run-against-all-the-files).
Or just try to make a commit and the pre-commit will be executed automatically.

## License

azure-operator is under the Apache 2.0 license. See the [LICENSE](LICENSE) file for
details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Add `global.podSecurityStandards.enforced` value for PSS migration.

## [8.2.0] - 2023-07-14

### Fixed

- Add necessary values for PSS policy warnings. 

## [8.1.0] - 2023-07-04

### Added

- Add support for customizing `controller-manager` `terminated-pod-gc-threshold` value through annotation `giantswarm.io/controllermanager-terminated-pod-gc-threshold`

### Removed

- Scripts: Remove `migrate-nginx-ingress-controller.sh`. ([#1729](https://github.com/giantswarm/azure-operator/pull/1729))

## [8.0.1] - 2023-05-16

### Fixed

- Fix helm values json schema.

## [8.0.0] - 2023-05-16

### Added

- Add new control-plane label to detect master nodes.

### Removed

- Remove logic that migrates CAPI CRDs from experimental group to new group.

### Changed

- Kubernetes 1.25 support.

## [7.2.0] - 2023-03-07

### Changed

- Bump k8s-api-healthz image to 0.2.0.

## [7.1.0] - 2023-02-07

### Changed

- Bump k8scc to support improve reliability of calico deployment process for new clusters.
- Enable `CronJobTimeZone` feature gate.

## [7.0.0] - 2023-01-11

### Changed

- Bump k8scc to support k8s 1.24.

## [6.0.3] - 2022-10-11

### Fixed

- Bump k8scc to improve calico crd installer startup time.

## [6.0.2] - 2022-09-01

### Fixed

- Ensure giantswarm user is always created first.

## [6.0.1] - 2022-08-31

### Fixed

- Bump k8scc to support cgroups v1 on containerd.

## [6.0.0] - 2022-08-31

### Fixed

- Bump `k8scc` to fix syntax error in `k8s-addons` script.
- Bump `k8scc` to enable authn and authz on `scheduler` and `controller-manager`.
- Enable external cloud controller manager in k8scc.

### Changed

- Bump `k8scc` to v14 to support kubernetes 1.23.
- Change default storage classes in order to use out-of-tree CSI provisioner.
- Improved storage class migration resource.

### Removed

- Remove --cloud-config flag from k8s components.

## [5.22.0] - 2022-07-04

### Changed

- Tighten pod and container security contexts for PSS restricted policies.

### Fixed

- Fix handling of `MachinePools'` status fields for empty node pools.

### Changed

- Bump `k8scc` to enable `auditd` monitoring for `execve` syscalls.

## [5.21.0] - 2022-06-22

### Changed

- Changes to EncryptionConfig in order to work with `encryption-provider-operator`.

### Fixed

- Add pause annotation before deleting old machinepool and azuremachinepool CRs during migration to non-exp.
- Update ownerReference UIDs during migration to non-exp.
- Avoid updating `AzureCluster` at every reconciliation loop in the `subnet` resource.
- Avoid saving `AzureCluster` status if there are no changes to avoid useless reconciliation loops.

## [5.20.0] - 2022-06-07

### Changed

- Bumped k8scc to latest version to fix `localhost` node name problem.

## [5.19.0] - 2022-06-07

### Added

- Added possibility to specify VNet CIDR in `AzureCluster`.
- Migrate MachinePool CRs from `exp.cluster.x-k8s.io/v1alpha3` to `cluster.x-k8s.io/v1beta1`
- Migrate AzureMachinePool CRs from `exp.infrastructure.cluster.x-k8s.io/v1alpha3` to `infrastructure.cluster.x-k8s.io/v1beta1`

### Changed

- Use systemd cgroup driver on masters and cgroups v2 worker nodes.
- Update github.com/Azure/azure-sdk-for-go to v58.1.0+incompatible
- Update github.com/giantswarm/apiextensions to v6.0.0
- Update github.com/giantswarm/certs to v4.0.0
- Update github.com/giantswarm/conditions to v0.5.0
- Update github.com/giantswarm/conditions-handler to v0.3.0
- Update github.com/giantswarm/k8sclient to v7.0.1
- Update github.com/giantswarm/k8scloudconfig to v13.4.0
- Update github.com/giantswarm/operatorkit to v7.0.1
- Update github.com/giantswarm/release-operator to v3.2.0
- Update github.com/giantswarm/tenantcluster to v6.0.0
- Update k8s.io/api to v0.22.2
- Update k8s.io/apiextensions-apiserver to v0.22.2
- Update k8s.io/apimachinery to v0.22.2
- Update k8s.io/client-go to v0.22.2
- Update sigs.k8s.io/cluster-api to v1.0.5
- Update sigs.k8s.io/cluster-api-provider-azure to v1.0.2
- Update sigs.k8s.io/controller-runtime to v0.10.3
- Bump various other dependencies to address CVEs.

### Fixed

- Set `AzureMachine.Status.Ready` according to AzureMachine's Ready condition.

## [5.18.0] - 2022-03-21

### Added

- Add VerticalPodAutoscaler CR.

## [5.17.0] - 2022-03-15

### Fixed

- Fix panic while checking for cgroups version during upgrade.

### Added

- Add GiantSwarmCluster tag to Vnet.

## [5.16.0] - 2022-02-23

### Changed

- Make nodepool nodes roll in case the user switches between cgroups v1 and v2.

## [5.15.0] - 2022-02-16

### Changed

- Drop dependency on `giantswarm/apiextensions/v2`.
- Bump `k8scloudconfig` to disable `rpc-statd`.

## [5.14.0] - 2022-02-02

### Added

- Add support for feature that enables forcing cgroups v1 for Flatcar version `3033.2.0` and above.

### Changed

- Upgraded to giantswarm/exporterkit v1.0.0
- Upgraded to giantswarm/microendpoint v1.0.0
- Upgraded to giantswarm/microkit v1.0.0
- Upgraded to giantswarm/micrologger v0.6.0
- Upgraded to giantswarm/versionbundle v1.0.0
- Upgraded to spf13/viper v1.10.0

## [5.13.0] - 2022-01-14

### Changed

- Bumped k8scc to latest version to support Calico 3.21.

## [5.12.0] - 2021-12-14

### Added

- Deal with AzureClusterConfig CR to avoid cluster operator conflict.

## [5.11.0] - 2021-12-10

### Removed

- Remove creation of legacy `AzureClusterConfig` CR as they are unused.

## [5.10.2] - 2021-12-07

### Fixed

- Consider case when API is down when checking if Master node is upgrading during node pool reconciliation.

## [5.10.1] - 2021-12-02

### Changed

- When looking for the encryption secret, search on all namespaces (to support latest cluster-operator).

## [5.10.0] - 2021-11-08

### Changed


client/factory.go
package client

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/Azure/azure-sdk-for-go/profiles/2019-03-01/authorization/mgmt/authorization"
	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
	"github.com/Azure/azure-sdk-for-go/services/dns/mgmt/2018-05-01/dns"
	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-11-01/network"
	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
	"github.com/Azure/go-autorest/autorest"
	"github.com/giantswarm/microerror"
	"github.com/giantswarm/micrologger"
	gocache "github.com/patrickmn/go-cache"

	"github.com/giantswarm/azure-operator/v8/pkg/credential"
	"github.com/giantswarm/azure-operator/v8/service/collector"
)

const (
	cacheHitLogKey       = "cacheHit"
	clientTypeLogKey     = "clientType"
	credentialNameLogKey = "credentialName" // nolint:gosec
	logLevelLogKey       = "level"
	logLevelDebug        = "debug"
	messageLogKey        = "message"
)

type FactoryConfig struct {
	AzureAPIMetrics    collector.AzureAPIMetrics
	CacheDuration      time.Duration
	CredentialProvider credential.Provider
	Logger             micrologger.Logger
}

// Factory is creating Azure clients for specified AzureConfig CRs, so basically for specified
// tenant clusters. All created clients are cached.
type Factory struct {
	credentialProvider credential.Provider
	logger             micrologger.Logger
	metricsCollector   collector.AzureAPIMetrics
	mutex              sync.Mutex

	// map [credentialName + client type] -> client
	cachedClients *gocache.Cache
}

type clientCreatorFunc func(autorest.Authorizer, collector.AzureAPIMetrics, string, string) (interface{}, error)

// NewFactory returns a new Azure client factory that is used throughout entire azure-operator
// lifetime.
func NewFactory(config FactoryConfig) (*Factory, error) {
	if config.AzureAPIMetrics == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.AzureAPIMetrics must not be empty", config)
	}
	if config.CacheDuration < 5*time.Minute { // cache at least for one reconciliation loop duration
		return nil, microerror.Maskf(invalidConfigError, "%T.CacheDuration must be at least 5 minutes", config)
	}
	if config.CredentialProvider == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.CredentialProvider must not be empty", config)
	}
	if config.Logger == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
	}

	factory := &Factory{
		logger:             config.Logger,
		credentialProvider: config.CredentialProvider,
		cachedClients:      gocache.New(config.CacheDuration, 2*config.CacheDuration),
		metricsCollector:   config.AzureAPIMetrics,
	}

	factory.cachedClients.OnEvicted(func(clientKey string, i interface{}) {
		factory.onEvicted(clientKey)
	})

	return factory, nil
}

// GetDeploymentsClient returns DeploymentsClient that is used for management of deployments and
// ARM templates. The client (for specified cluster) is cached after creation, so the same client
// is returned every time.
func (f *Factory) GetDeploymentsClient(credentialNamespace, credentialName string) (*resources.DeploymentsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "DeploymentsClient", newDeploymentsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toDeploymentsClient(client), nil
}

// GetDisksClient returns DisksClient that is used for management of virtual disks.
// The client (for specified cluster) is cached after creation, so the same client
// is returned every time.
func (f *Factory) GetDisksClient(credentialNamespace, credentialName string) (*compute.DisksClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "DisksClient", newDisksClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toDisksClient(client), nil
}

// GetGroupsClient returns GroupsClient that is used for management of resource groups for the
// specified cluster. The created client is cached for the time period specified in the factory
// config.
func (f *Factory) GetGroupsClient(credentialNamespace, credentialName string) (*resources.GroupsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "GroupsClient", newGroupsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toGroupsClient(client), nil
}

// GetInterfacesClient returns InterfacesClient that is used for management of network interfaces for the
// specified cluster. The created client is cached for the time period specified in the factory
// config.
func (f *Factory) GetInterfacesClient(credentialNamespace, credentialName string) (*network.InterfacesClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "InterfacesClient", newInterfacesClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toInterfacesClient(client), nil
}

// GetDNSRecordSetsClient returns RecordSetsClient that is used for management of DNS records.
// The client (for specified cluster) is cached after creation, so the same client
// is returned every time.
func (f *Factory) GetDNSRecordSetsClient(credentialNamespace, credentialName string) (*dns.RecordSetsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "RecordSetsClient", newDNSRecordSetsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toDNSRecordSetsClient(client), nil
}

// GetVirtualMachineScaleSetsClient returns VirtualMachineScaleSetsClient that is used for
// management of virtual machine scale sets for the specified cluster. The created client is cached
// for the time period specified in the factory config.
func (f *Factory) GetVirtualMachineScaleSetsClient(credentialNamespace, credentialName string) (*compute.VirtualMachineScaleSetsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualMachineScaleSetsClient", newVirtualMachineScaleSetsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualMachineScaleSetsClient(client), nil
}

// GetVirtualMachineScaleSetVMsClient returns GetVirtualMachineScaleSetVMsClient that is used for
// management of virtual machine scale set instances for the specified cluster. The created client
// is cached for the time period specified in the factory config.
func (f *Factory) GetVirtualMachineScaleSetVMsClient(credentialNamespace, credentialName string) (*compute.VirtualMachineScaleSetVMsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualMachineScaleSetVMsClient", newVirtualMachineScaleSetVMsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualMachineScaleSetVMsClient(client), nil
}

// GetVirtualNetworksClient returns *network.VirtualNetworksClient that is used for management of Azure
// virtual networks. The created client is cached for the time period specified in the factory config.
func (f *Factory) GetVirtualNetworksClient(credentialNamespace, credentialName string) (*network.VirtualNetworksClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualNetworksClient", newVirtualNetworksClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualNetworksClient(client), nil
}

// GetSnapshotsClient returns *compute.SnapshotsClient that is used for management of Azure
// snapshots (both disks and VMs). The created client is cached for the time period
// specified in the factory config.
func (f *Factory) GetSnapshotsClient(credentialNamespace, credentialName string) (*compute.SnapshotsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "SnapshotsClient", newSnapshotsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toSnapshotsClient(client), nil
}

// GetStorageAccountsClient returns *storage.AccountsClient that is used for management of Azure
// storage accounts for the specified cluster. The created client is cached for the time period
// specified in the factory config.
func (f *Factory) GetStorageAccountsClient(credentialNamespace, credentialName string) (*storage.AccountsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "StorageAccountsClient", newStorageAccountsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toStorageAccountsClient(client), nil
}

// GetSubnetsClient returns *network.SubnetsClient that is used for management of Azure subnets.
// The created client is cached for the time period specified in the factory config.
func (f *Factory) GetSubnetsClient(credentialNamespace, credentialName string) (*network.SubnetsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "SubnetsClient", newSubnetsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toSubnetsClient(client), nil
}

// GetNatGatewaysClient returns *network.NatGatewaysClient that is used for management of Nat Gateways.
// The created client is cached for the time period specified in the factory config.
func (f *Factory) GetNatGatewaysClient(credentialNamespace, credentialName string) (*network.NatGatewaysClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "NatGatewaysClient", newNatGatewaysClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toNatGatewaysClient(client), nil
}

// GetNetworkSecurityGroupsClient returns *network.SecurityGroupsClient that is used for management of Network Security Groups.
// The created client is cached for the time period specified in the factory config.
func (f *Factory) GetNetworkSecurityGroupsClient(credentialNamespace, credentialName string) (*network.SecurityGroupsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "NetworkSecurityGroupsClient", newNetworkSecurityGroupsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toNetworkSecurityGroupsClient(client), nil
}

func (f *Factory) GetPublicIPAddressesClient(credentialNamespace, credentialName string) (*network.PublicIPAddressesClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "PublicIPAddressesClient", newPublicIPAddressesClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toPublicIPAddressesClient(client), nil
}

// GetResourceSkusClient returns *compute.ResourceSkusClient that is used for reading VM instance types.
// The created client is cached for the time period specified in the factory config.
func (f *Factory) GetResourceSkusClient(credentialNamespace, credentialName string) (*compute.ResourceSkusClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "ResourceSkusClient", newResourceSkusClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toResourceSkusClient(client), nil
}

func (f *Factory) GetVirtualNetworkPeeringsClient(credentialNamespace, credentialName string) (*network.VirtualNetworkPeeringsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualNetworkPeeringsClient", newVnetPeeringClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualNetworkPeeringsClient(client), nil
}

func (f *Factory) GetVirtualNetworkGatewaysClient(credentialNamespace, credentialName string) (*network.VirtualNetworkGatewaysClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualNetworkGatewaysClient", newVirtualNetworkGatewaysClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualNetworkGatewaysClient(client), nil
}

func (f *Factory) GetVirtualNetworkGatewayConnectionsClient(credentialNamespace, credentialName string) (*network.VirtualNetworkGatewayConnectionsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "VirtualNetworkGatewayConnectionsClient", newVirtualNetworkGatewayConnectionsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toVirtualNetworkGatewayConnectionsClient(client), nil
}

func (f *Factory) GetRoleAssignmentsClient(credentialNamespace, credentialName string) (*authorization.RoleAssignmentsClient, error) {
	client, err := f.getClient(credentialNamespace, credentialName, "RoleAssignmentsClient", newRoleAssignmentsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toRoleAssignmentsClient(client), nil
}

func (f *Factory) getClient(credentialNamespace, credentialName string, clientType string, createClient clientCreatorFunc) (interface{}, error) {
	l := f.logger.With(
		logLevelLogKey, logLevelDebug,
		messageLogKey, "get client",
		credentialNameLogKey, credentialName,
		clientTypeLogKey, clientType)

	clientKey := getClientKey(credentialName, clientType)
	var client interface{}
	f.mutex.Lock()
	defer f.mutex.Unlock()

	if cachedClient, ok := f.cachedClients.Get(clientKey); ok {
		// client found, it will be refreshed in cache
		l.Log(cacheHitLogKey, true)
		client = cachedClient
	} else {
		// client not found, create it, it will be saved in cache
		l.Log(cacheHitLogKey, false)
		newClient, err := f.createClient(credentialNamespace, credentialName, createClient)
		if err != nil {
			return nil, microerror.Mask(err)
		}
		client = newClient
	}

	// refresh existing client or set new one
	f.cachedClients.SetDefault(clientKey, client)
	return client, nil
}

func (f *Factory) createClient(credentialNamespace, credentialName string, createClient clientCreatorFunc) (interface{}, error) {
	organizationCredentialsConfig, subscriptionID, partnerID, err := f.credentialProvider.GetOrganizationAzureCredentials(context.Background(), credentialNamespace, credentialName)
	if err != nil {
		return nil, microerror.Mask(err)
	}
	authorizer, err := organizationCredentialsConfig.Authorizer()
	if err != nil {
		return nil, microerror.Mask(err)
	}
	client, err := createClient(authorizer, f.metricsCollector, subscriptionID, partnerID)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return client, nil
}

func getClientKey(credentialName string, clientType string) string {
	return fmt.Sprintf("%s.%s", credentialName, clientType)
}

func getClientKeyParts(clientKey string) (credentialName, clientType string) {
	parts := strings.Split(clientKey, ".")
	partsCount := len(parts)

	if partsCount > 2 {
		credentialName = strings.Join(parts[0:partsCount-1], ".")
		clientType = parts[partsCount-1]
	} else if partsCount == 2 {
		credentialName, clientType = parts[0], parts[1]
	} else {
		// this should never happen, don't return error, this is for logging only
		credentialName, clientType = "unknown", "unknown"
	}

	return
}

func (f *Factory) onEvicted(clientKey string) {
	credentialName, clientType := getClientKeyParts(clientKey)
	f.logger.Log(
		logLevelLogKey, logLevelDebug,
		messageLogKey, "client evicted",
		credentialNameLogKey, credentialName,
		clientTypeLogKey, clientType)
}
